3a.train_individual_models/nbconverted/4.apply_models_to_compounds.py

```diff
@@ -0,0 +1,153 @@
+all_predictions = []
+
+for plate, items in model_paths_dir.items():
+    print(f"Processing model: {plate}")
+    # Check if the model paths exist
+    if not items["final_model"] or not items["shuffled_model"]:
+        print(f"Model paths for {plate} do not exist.")
+        continue
+
+    # Load the models
+    final_model = load(items["final_model"][0])
+    shuffled_model = load(items["shuffled_model"][0])
+
+    # Load the normalized data
+    if plate == "combined_batch1":
+        # Concatenate all plate paths for the combined case
+        norm_data_frames = [pd.read_parquet(path) for path in items["norm_path"]]
+        norm_data = pd.concat(norm_data_frames, ignore_index=True)
+    else:
+        # Load the single parquet file for non-combined plates (same plate as model)
+        norm_path = items["norm_path"][0]
+        norm_data = pd.read_parquet(norm_path)
+
+    # Filter norm_data for only the rows with compound treatment (remove DMSO)
+    norm_data = norm_data[norm_data["Metadata_treatment"] != "DMSO"]
+
+    # Get the feature columns used in the model to filter the data
+    model_columns = final_model.feature_names_in_  # Feature names are same in shuffled model
+
+    # Filter the data to only include the model columns including metadata columns
+    meta_cols = norm_data.columns[norm_data.columns.str.startswith("Metadata_")]
+    norm_data = norm_data[list(meta_cols) + list(model_columns)]
+
+    # Drop any rows with NaN in the model columns after filtering
+    norm_data = norm_data.dropna(subset=model_columns)
+
+    # Get X and y data
+    X, y = get_X_y_data(norm_data, label="Metadata_cell_type")
+
+    # Make predictions (probabilities)
+    y_pred_final_probs = final_model.predict_proba(X)
+    y_pred_shuffled_probs = shuffled_model.predict_proba(X)
+
+    # Add predicted probabilities back to the metadata columns
+    predictions_df = norm_data[list(meta_cols)].copy()  # Copy metadata columns
+
+    # Create DataFrames for final and shuffled model predictions
+    final_predictions_df = predictions_df.copy()
+    final_predictions_df["predicted_probas"] = y_pred_final_probs[:, 1]  # Probability of the positive class (healthy)
+    final_predictions_df["model_type"] = "final"
+
+    shuffled_predictions_df = predictions_df.copy()
+    shuffled_predictions_df["predicted_probas"] = y_pred_shuffled_probs[:, 1]  # Probability of the positive class (healthy)
+    shuffled_predictions_df["model_type"] = "shuffled"
+
+    # Concatenate the two DataFrames
+    predictions_df = pd.concat([final_predictions_df, shuffled_predictions_df], ignore_index=True)
+
+    # Add the model_name column
+    predictions_df["model_name"] = plate
+
+    # Append the DataFrame to the list
+    all_predictions.append(predictions_df)
+
+# Concatenate all predictions into a single DataFrame
+all_predictions_df = pd.concat(all_predictions, ignore_index=True)
+
+# Prior to saving, fix Pathways bug for two compounds
```

Comment:
what do you mean `bug`? Is it a typo?

Should we correct this at the source so we don't have to in the future?

Reply:
We definitely will need to correct at the source, it was just out-of-scope for this PR. I have made it into an issue (#16), likely Erik or I can address it. I couldn't see anything obvious for why there is this problem with first glance at the code, so it might take some time to diagnose. For now, this is a bandaid.